### ISSUE TITLE
Provide support for an ordered list of search predicates when matching known build issues

### ISSUE THREAD

**Issue by ericstj** [dotnet/dnceng#981]
When crafting a match to identify a failing test or build issue, it's sometimes difficult to create a single-line match that uniquely identifies an issue.

Take an exception message for example -- I don't want to match on that exception message whenever it happens.  I might want to only match on it when it occurs in a particular test, or only with some particular method on the stack.  I still might care about the message *AND* the stack or test.

Today it's not possible to do this, since I can only define a single match and that match can only span a single line.

I don't think multi-line matches help - since those cost more in overall performance and are harder to write correctly.

Instead, I propose we add a conditional logical AND of predicates.  This could be represented with an array of predicates.

For example:
```json
{
  "ErrorMessage": [ "RequestDuration_EnrichmentHandler_Success_Recorded", "System.Net.Quic.QuicException : Stream aborted by peer" ]
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

To match, the service could simply first match the first string then only proceed matching subsequent strings if it found a match for the first in the same source document.  Each entry in the array would only be a single line match. This would be a conditional logical AND (&&), where it short-circuits, or stops evaluation as soon as a false clause is encountered.  We could guide folks to start with the most constraining clause as their first clause to minimize matching (and perhaps even make it an error and halt if the first clause matched too many things). 
 

### Release Note Category
- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements
### Release Note Description
Known Issues now support a list of errors. Check out the [documentation](https://github.com/dotnet/arcade/blob/c78ffee889131fcdc4c68968ee0f67fe9ca6339b/Documentation/Projects/Build%20Analysis/KnownIssues.md#fill-out-known-issues-with-a-list-of-errors) for more information.

**Comment by hoyosjs** [comment]
Question: 
- What to do when both error message and error pattern appear?
- Is this just literal matches?

**Comment by ericstj** [comment]
I'm not proposing any change to the behavior of `ErrorMessage` vs `ErrorPattern`, you still need to use one or the other.  It will be invalid to specify both, just as it is today.

I think the conditional AND approach should work fine for both message and pattern.  You could also have an array for `ErrorPattern` that serves as a conditional AND of those RegEx matches:
```json
{
  "ErrorMessage": ""
  "ErrorPattern":  [ "RequestDuration_EnrichmentHandler_.*", "System\\.Net.Quic\\.QuicException : Stream aborted by peer \\((268|58|267)\\)" ],
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

**Comment by ericstj** [comment]
Note: the array here need not be the user experience if folks prefer something different, that was just what seemed the most straight-forward to me to represent this concept.  I could also imagine us adding query object or something more advanced.

**Comment by agocke** [comment]
Is this faster than multi-line regex? I'd be somewhat surprised

**Comment by ericstj** [comment]
I'm not very familiar with why we haven't enabled multi-line regexes and the cost associated.  It could have to do with how things are indexed.

I suggested this because it seems like a simple alternative that still achieves the need of matching multiple lines and might be easier/lest-costly to implement and potentially easier to use that multi-line regex.

**Comment by ericstj** [comment]
Testing both myself (to craft a multiline regex) and Regex itself.

```
|                Method |      Mean |     Error |    StdDev |
|---------------------- |----------:|----------:|----------:|
| ConditionalAndMessage |  1.380 us | 0.0100 us | 0.0093 us |
|   ConditionalAndRegex |  4.140 us | 0.0137 us | 0.0115 us |
|        MultiLineRegex | 69.408 us | 0.1993 us | 0.1865 us |
```

This is based on trying to create a match for https://[helixre107v0xd1eu3ibi6ka.blob.core.windows.net/dotnet-runtime-refs-pull-93120-merge-47197dce91b04dbf87/System.Net.Http.Functional.Tests/3/console.e4e6d7cd.log?helixlogtype=result](https://helixre107v0xd1eu3ibi6ka.blob.core.windows.net/dotnet-runtime-refs-pull-93120-merge-47197dce91b04dbf87/System.Net.Http.Functional.Tests/3/console.e4e6d7cd.log?helixlogtype=result)

Here's the benchmark: https://gist.github.com/ericstj/56d847c2989fa459736f774f82447d83

**Comment by AlitzelMendez** [comment]
Hi @ericstj,

I started working on this briefly and created this draft pr: https://dev.azure.com/dnceng/internal/_git/dotnet-helix-service/pullrequest/35983

I just want to double check that we are on the same page about the expectation of this feature request.

This would be a simulation of multiline, in where all the lines that should be included should be listed. From our side we will review the file as usual, line by line, but the regex/message that we are trying to match could change in every line, depending if the previous line was matched.

For example, if we have the following lines in a file:
A
B
C

and the error message is [B,C]
we are going to review the file in the following way: A != D / B = B / C = C  -> (there is a match)
In other words we will look for B until we find it and the if there is a B we will look for C on the following line 

but for example if we have this lines and same message  [B,C]
A
B
D
C
we are going to review the file in the following way: A != B / B = B /  D != C / C  != B   (not a match)  
In this case as there wasn't a match for C we started looking for B again

is this the expectation??

**Comment by ericstj** [comment]
>  if there is a B we will look for C on the following line

I was imagining that there wouldn't necessarily be an ordering constraint to the search, but instead it behaved like multiple clauses that all must be true.

So if I give it the clause [B, C], I would want the service to find all logs that contain both B and C, regardless of where they appear in the log.  Maybe the algorithm is first find a log that contains B, then check that it also contains C. 

Of course this is just one type of search operator.  I could imagine variations -- where C must be the on the following line as you suggest, or where C can be on any line following B.  Those seem like useful specializations, but I think the general unordered AND is more useful overall.

**Comment by AlitzelMendez** [comment]
The general unordered case puts us in a more performance-expensive situation, in which we need to review every line against multiple messages/regexes, affecting our performance depending on how many messages were shared. We are currently reviewing around 150 known issues, so this would add more burden to this and it seems very similar to having multi-line regexes, which are costly for us to support. I will check if this is something we can support performance-wise.

The unordered scenario is cheaper in the sense that we don’t need to add more checks, we just need to change which message we are reviewing against the line. My only concern about not including every line is that we might start matching very unexpected things, especially for helix logs. For example, “name of a test” and “failed” could appear anywhere on a helix log and a “failed” message could belong to any test, but if we are just checking if those two lines are part of the log, we will make a match. Other than that, this is a route we can totally take.

In conclusion, let me look at the data and decide if we can go for the unordered route. If not, let’s discuss more about the ordered one and whether it would be actually helpful or not.

**Comment by ericstj** [comment]
I think the ordered implementation would still work for the scenarios we've seen.  All the cases I have seen would have some order of logged messages.  Like some test starts - then process crashes -- or some test starts, an exception is thrown with a generic message, then a specific method is on the stack.

It wasn't what I was thinking of when I suggested this feature - since I was imaging more the way literal searches work in search engines - but I think it could work if we can train folks how to use it.

The only situation I can think of where a failure signature might be unordered would be some failure mode that has to do with multi-threading.  Like test A and test F run concurrently - where either starts first- and some failure is hit.  I don't think I've seen a failure mode like this where I could also nail down what might be significant to represent a search query (if I could do that, I'm probably pretty close to fixing the issue...).  So while I think an unordered search might be a more intuitive UX - I can see how ordered might be better here.

**Comment by AlitzelMendez** [comment]
Thank you @ericstj, the ordered implementation works a lot better for us. I will proceed with this, taking into consideration what you mentioned about lines not being consecutive. I will also make sure to be clear in the documentation that we are reviewing the lines in an ordered way.

**Comment by AlitzelMendez** [comment]
PR: https://dev.azure.com/dnceng/internal/_git/dotnet-helix-service/pullrequest/35983
